spec/models/cost_query/cache_spec.rb

```diff
@@ -0,0 +1,126 @@
+#-- copyright
+# OpenProject Reporting Plugin
+#
+# Copyright (C) 2010 - 2014 the OpenProject Foundation (OPF)
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# version 3.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#++
+
+require 'spec_helper'
+require File.join(File.dirname(__FILE__), '..', '..', 'support', 'configuration_helper')
+
+describe CostQuery::Cache do
+  include OpenProject::Reporting::SpecHelper::ConfigurationHelper
+
+  def all_caches
+    [ CostQuery::GroupBy::CustomFieldEntries,
+      CostQuery::GroupBy,
+      CostQuery::Filter::CustomFieldEntries,
+      CostQuery::Filter ]
+  end
+
+  def expect_reset_on_caches
+    all_caches.each do |klass|
+      expect(klass).to receive(:reset!)
+    end
+  end
+
+  def expect_no_reset_on_caches
+    all_caches.each do |klass|
+      expect(klass).to_not receive(:reset!)
+    end
+  end
+
+  def reset_cache_keys
+    # resetting internal caching keys to avoid dependencies with other specs
+    described_class.send(:custom_fields_updated_on=, nil)
+    described_class.send(:custom_fields_id_sum=, 0)
+  end
+
+  def custom_fields_exist
+    allow(WorkPackageCustomField).to receive(:maximum).and_return(Time.now)
+    allow(WorkPackageCustomField).to receive(:sum).and_return(42)
+    allow(WorkPackageCustomField).to receive(:count).and_return(23)
+  end
+
+  def no_custom_fields_exist
+    allow(WorkPackageCustomField).to receive(:maximum).and_return(nil)
+    allow(WorkPackageCustomField).to receive(:sum).and_return(0)
+    allow(WorkPackageCustomField).to receive(:count).and_return(0)
+  end
```

Comment:
I am not sure if I like you to mock so heavily regarding basic functions of `WorkPackageCustomField`.

This will surely make for green tests, but it is so whitebox-ish, that I could barely use those specs for refactoring. Not sure if this is your intention, though.